AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/src/ApplyMutation.dfy

```diff
@@ -130,44 +135,15 @@ module {:options "/functionSyntax:4" } InternalApplyMutation {
             input.keyManagerStrategy.Modifies,
             input.SystemKey.Modifies
   {
-    // -= Fetch Commitment and Index
     var storage := input.storage;
     var keyManagerStrategy := input.keyManagerStrategy;
     var SystemKey := input.SystemKey;
     // var logicalKeyStoreName := input.logicalKeyStoreName;
-    var fetchMutation? := storage.GetMutation(
-      Types.AwsCryptographyKeyStoreTypes.GetMutationInput(
-        Identifier := input.MutationToken.Identifier));
-    var fetchMutation: KeyStoreTypes.GetMutationOutput :- fetchMutation?
-    .MapFailure(e => Types.Error.AwsCryptographyKeyStore(e));
+    var commitmentAndIndex :- CommitmentAndIndex.FetchMutationItems(storage, input.MutationToken.Identifier);
+    :- CommitmentAndIndex.TokenAndMutationItemsAggree(commitmentAndIndex, input.MutationToken);
+    :- CommitmentAndIndex.MutationItemsValidUTF8(commitmentAndIndex);
 
-      // -= Validate Commitment and Index
-    :- Need(
-      fetchMutation.MutationCommitment.Some?,
-      Types.MutationInvalidException(
-        message := "No Mutation is in-flight for this Branch Key ID " + input.MutationToken.Identifier + " ."
-      ));
-    :- Need(
-      input.MutationToken.UUID == fetchMutation.MutationCommitment.value.UUID,
-      Types.MutationInvalidException(
-        message := "The Token and the Mutation Commitment read from storage disagree."
-        + " This indicates that the Token is for a different Mutation than the one in-flight."
-        + " A possible cause is this token is from an earlier Mutation that already finished?"
-        + " Branch Key ID: " + input.MutationToken.Identifier + ";"
-        + " Mutation Commitment UUID: " + fetchMutation.MutationCommitment.value.UUID + ";"
-        + " Token UUID: " + input.MutationToken.UUID + ";"
-      ));
-    :- Need(
-      fetchMutation.MutationIndex.Some?,
-      Types.MutationInvalidException(
-        message := "No Mutation Index exsists for this in-flight mutation of Branch Key ID " + input.MutationToken.Identifier + " ."
-      ));
-    var CommitmentAndIndex :- Mutations.ValidateCommitmentAndIndexStructures(
-      input.MutationToken,
-      fetchMutation.MutationCommitment.value,
-      fetchMutation.MutationIndex.value);
-
-    var commitmentIsVerified :- SystemKeyHandler.VerifyCommitment(CommitmentAndIndex.Commitment, SystemKey);
+    var commitmentIsVerified :- SystemKeyHandler.VerifyCommitment(commitmentAndIndex.commitment, SystemKey);
```

Comment:
Should we validate `token.Identifier == index.Identifier`?

Reply:
We can add an assert if you would like.  

`CommitmentAndIndex.FetchMutationItems` ensures the two items have the same identifiers (bkid and uuid),
while `CommitmentAndIndex.TokenAndMutationItemsAggree` ensures the three have the same identifiers (bkid and uuid).
But I would rather not have another runtime check.
Formal verification is here to help us ensure we have sufficient checks,  
but can also be used to help us identify redundant checks. 

We can also review the post conditions on `SystemKeyHandler.VerifyCommitment`;
it should ensure that the IDs and content are consistent b/w the input and output.